### PR TITLE
[BEAM-4130] Add job submission capabilities to Flink runner.

### PR DESCRIPTION
**Pull request by robertwb**
This is a rebase of the submission service in the hacking branch, updated to reflect the new state of artifact staging. 

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @axelmagn, @angoenka Please take a look.

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/InMemoryJobService.java [hunk 1]
```diff
@@ -60,17 +60,28 @@
 
   public static InMemoryJobService create(
       Endpoints.ApiServiceDescriptor stagingServiceDescriptor, JobInvoker invoker) {
-    return new InMemoryJobService(stagingServiceDescriptor, invoker);
+    return new InMemoryJobService(stagingServiceDescriptor, (String session) -> "token", invoker);
+  }
+
+  public static InMemoryJobService create(
+      Endpoints.ApiServiceDescriptor stagingServiceDescriptor,
+      Function<String, String> stagingServiceTokenProvider,
```
Comment (axelmagn) [inline runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/InMemoryJobService.java]: maybe document that this function expects a preparation ID as input.

Comment (robertwb) [reply runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/InMemoryJobService.java]: Good point.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvocation.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    this.executorService = executorService;
+    this.pipeline = pipeline;
+    this.pipelineOptions = pipelineOptions;
+    this.invocationFuture = null;
+    this.jobState = Enum.STOPPED;
+    this.stateObservers = new ArrayList<>();
+  }
+
+  private PipelineResult runPipeline() throws Exception {
+    LOG.trace("Translating pipeline from proto");
```
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvocation.java]: Done.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvocation.java [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    return FlinkRunner.createPipelineResult(result, pipelineOptions);
+  }
+
+  @Override
+  public void start() {
+    LOG.trace("Starting job invocation {}", getId());
+    synchronized (this) {
+      setState(Enum.STARTING);
+      invocationFuture = executorService.submit(this::runPipeline);
+      setState(Enum.RUNNING);
```
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvocation.java]: We could defer it until after the pipeline translation, but that doesn't seem to be a significant difference, and it's cleaner to have all the state transitions locally here. I added a TODO to actually do this right (which we can't do from runPipeline alone).

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvocation.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+      Futures.addCallback(
+          invocationFuture,
+          new FutureCallback<PipelineResult>() {
+            @Override
+            public void onSuccess(
+                @Nullable PipelineResult pipelineResult) {
```
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvocation.java]: Done.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvocation.java [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+  }
+
+  @Override
+  public String getId() {
+    return id;
+  }
+
+  @Override
+  public void cancel() {
+    LOG.trace("Canceling job invocation {}", getId());
```
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvocation.java]: Done.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvocation.java [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    synchronized (this) {
```
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvocation.java]: Done.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvocation.java [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+      if (this.invocationFuture != null) {
+        this.invocationFuture.cancel(true /* mayInterruptIfRunning */);
```
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvocation.java]: Done.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvocation.java [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+      }
+    }
+  }
+
+  @Override
+  public synchronized Enum getState() {
+    return this.jobState;
```
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvocation.java]: jobState has a longer lifetime.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvocation.java [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+
+  @Override
+  public synchronized void addMessageListener(Consumer<JobMessage> messageStreamObserver) {
+    LOG.warn("addMessageObserver() not yet implemented.");
+  }
+
+  private synchronized void setState(Enum state) {
+    this.jobState = state;
+    for (Consumer<Enum> observer : stateObservers) {
+      observer.accept(state);
```
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvocation.java]: This also ensures that observers always receive state changes in order.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvocation.java [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+    }
+  }
+
+  /** Indicates whether the given pipeline has any unbounded PCollections. */
+  private static boolean hasUnboundedPCollections(RunnerApi.Pipeline pipeline) {
+    checkArgument(pipeline != null);
```
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvocation.java]: Done.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvoker.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+  private FlinkJobInvoker(ListeningExecutorService executorService) {
+    this.executorService = executorService;
+  }
+
+  @Override
+  public JobInvocation invoke(
+      RunnerApi.Pipeline pipeline, Struct options, @Nullable String artifactToken)
+      throws IOException {
+    // TODO: How to make Java/Python agree on names of keys and their values?
+    LOG.trace("Parsing pipeline options");
```
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvoker.java]: There isn't really anything to log here. As long as options parse correctly, we'll be logging soon enough.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvoker.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    FlinkPipelineOptions flinkOptions = PipelineOptionsTranslation.fromProto(options)
+        .as(FlinkPipelineOptions.class);
+
+    String invocationId = String.format(
+        "%s_%s", flinkOptions.getJobName(), UUID.randomUUID().toString());
+    LOG.debug("Invoking job {}", invocationId);
```
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvoker.java]: Done.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvoker.java [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+    // Set Flink Master to [auto] if no option was specified.
+    if (flinkOptions.getFlinkMaster() == null) {
+      flinkOptions.setFlinkMaster("[auto]");
+    }
+
+    flinkOptions.setRunner(null);
```
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvoker.java]: @axelmagn ?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobServerDriver.java [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+    private String artifactStagingPath = "/tmp/beam-artifact-staging";
+  }
+
+  public static void main(String[] args) {
+    ServerConfiguration configuration = new ServerConfiguration();
+    CmdLineParser parser = new CmdLineParser(configuration);
+    try {
+      parser.parseArgument(args);
+    } catch (CmdLineException e) {
+      e.printStackTrace(System.err);
```
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobServerDriver.java]: Done.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobServerDriver.java [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+  }
+
+  public static FlinkJobServerDriver create(
+      ServerConfiguration configuration,
+      ListeningExecutorService executor,
+      ServerFactory serverFactory) {
+    return new FlinkJobServerDriver(configuration, executor, serverFactory);
+  }
+
+  private final ListeningExecutorService executor;
```
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobServerDriver.java]: Done.

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/artifact/BeamFileSystemArtifactSource.java [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+    }
+    throw new IllegalArgumentException("No such artifact: " + name);
+  }
+
+  private ArtifactApi.ProxyManifest getProxyManifest() throws IOException {
+    if (proxyManifest == null) {
+      ArtifactApi.ProxyManifest.Builder builder = ArtifactApi.ProxyManifest.newBuilder();
+      JsonFormat.parser().merge(Channels.newReader(
+          FileSystems.open(FileSystems.matchNewResource(retrievalToken, false /* is directory */)),
+          StandardCharsets.UTF_8.name()), builder);
```
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/artifact/BeamFileSystemArtifactSource.java]: Yes, just like we're standardizing on Json (and Json is always supposed to be UTF-8.)

File: runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/artifact/BeamFileSystemArtifactSourceTest.java [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+
+          @Override public void onError(Throwable throwable) {
+            throw new RuntimeException(throwable);
+          }
+
+          @Override public void onCompleted() {
+          }
+        });
+
+    return stagingTokenHolder[0];
```
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/artifact/BeamFileSystemArtifactSourceTest.java]: This is a direct call, nothing asynchronous to wait on.

File: runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/artifact/BeamFileSystemArtifactSourceTest.java [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+      }
+
+      @Override public void onError(Throwable throwable) {
+        throw new RuntimeException(throwable);
+      }
+
+      @Override public void onCompleted() {
+      }
+    });
+    return contents.toString();
```
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/artifact/BeamFileSystemArtifactSourceTest.java]: Same.

**Comment by angoenka** [comment]
LGTM
Please make sure that the checks are green.

**Comment by angoenka** [comment]
@robertwb Based on the discussion in the chat thread, should we remove BeamFileSystemArtifactSource as ArtifactRetrivalService is meant to do the same job.
If you want we can keep it as it is as I can clean it up when I go about removing ArtifactSource interface.

**Comment by robertwb** [comment]
It's just hitting org.apache.beam.runners.direct.portable.ReferenceRunnerTest.pipelineExecution which is unrelated and flaky in head too.

**Comment by angoenka** [comment]
In that case we can go ahead with the merge.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvoker.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (kennknowles) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvoker.java:66]: I came across this today as well. Do you know?

[the remark this answers is not in the document]
Comment (robertwb) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvoker.java:66]: I don't remember.

**Pull request merged**